### PR TITLE
Move targetCPUPercentile into a flag

### LINKED ISSUE #4786: Support for different TargetCPUPercentiles through alternative recommender
It would be nice if one could choose to use different percentiles as base for VPA recommendation. We can do that by:

1. Adding flags so one can run VPA recommender that sets recommendations on VPA Objects with a specified name. (and also if it should set recommendation for VPA objects that don't specify which recommender should generate recommendation for them)

Currently VPA recommender sets recommendations  for VPAs which ([source](https://github.com/kubernetes/autoscaler/blob/b9182834cb3f0769058cd9dc59daaf40641e19f6/vertical-pod-autoscaler/pkg/recommender/input/cluster_feeder.go#L353)):
* don't specify which recommender they use or 
* are explicitly configured to use `default` recommender.

2. Adding flags so one can change target CPU percentiles.
3. Adding instruction on how to run two recommenders with different configuration in one cluster.

I'll check if it's feasible to add e2e test for running two VPA recommenders at the same time.

### PR DESCRIPTION
**Pull request by matthyx**
#### Which component this PR applies to?
vertical-pod-autoscaler

#### What type of PR is this?
/kind feature

#### Which issue(s) this PR fixes:

Fixes #4786

#### Does this PR introduce a user-facing change?

```release-note
Add flags to specify recommender name and targetCPUPercentile via command line arguments.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by matthyx** [comment]
@jbartosik something like that?
Let me know and I'll add some test coverage and docs.

File: vertical-pod-autoscaler/pkg/recommender/main.go
Comment (jbartosik) [inline vertical-pod-autoscaler/pkg/recommender/main.go:36]: I think we can drop this const, set `"default"` as the default value of `recommenderName` flag

**Review by jbartosik — CHANGES_REQUESTED** [review]
Thank you, some comments

File: vertical-pod-autoscaler/pkg/recommender/input/cluster_feeder.go
Comment (matthyx) [reply vertical-pod-autoscaler/pkg/recommender/input/cluster_feeder.go]: Can you check if I didn't mess up the logic here?

**Comment by matthyx** [comment]
/remove-lifecycle stale
@jbartosik can check and tell me what's missing, please?

File: vertical-pod-autoscaler/pkg/recommender/input/cluster_feeder.go
Comment (jbartosik) [inline vertical-pod-autoscaler/pkg/recommender/input/cluster_feeder.go:61]: This comment seems incorrect. How about

// If recommender name is equal to DefaultRecommenderName it will handle VPA object which don't specify recommender name explicitly (and so implicitly specify that the default recommender should handle them)

File: vertical-pod-autoscaler/pkg/recommender/logic/recommender.go
Comment (jbartosik) [inline vertical-pod-autoscaler/pkg/recommender/logic/recommender.go:29]: I think this flag could use a more detailed description. I'll write my idea later.

**Review by jbartosik — CHANGES_REQUESTED** [review]
Thanks. Mostly looking good, I made a few, hopefully small suggestions.

File: vertical-pod-autoscaler/pkg/recommender/logic/recommender.go
Comment (jbartosik) [reply vertical-pod-autoscaler/pkg/recommender/logic/recommender.go:29]: What do you think about the following description for the flag:

CPU usage percentile that will be used as a base for CPU target recommendation. Doesn't affect CPU lower bound, CPU upper bound nor memory recommendations.

?

File: vertical-pod-autoscaler/pkg/recommender/main.go
Comment (jbartosik) [reply vertical-pod-autoscaler/pkg/recommender/main.go:37]: What do you think about the following description for the flag:

Set the recommender name. Recommender will generate recommendations for VPAs that configure the same recommender name. If the recommender name is left as default it will also generate recommendations that don't explicitly specify recommender. You shouldn't run two recommenders with the same name in a cluster.

?

**Review by jbartosik — CHANGES_REQUESTED** [review]

File: vertical-pod-autoscaler/pkg/recommender/input/cluster_feeder.go
Comment (matthyx) [reply vertical-pod-autoscaler/pkg/recommender/input/cluster_feeder.go:363]: I think you're right... but unfortunately I find no "nicer" way of implementing it. Complexity goes to the sky.

File: vertical-pod-autoscaler/pkg/recommender/input/cluster_feeder_test.go
Comment (matthyx) [reply vertical-pod-autoscaler/pkg/recommender/input/cluster_feeder_test.go:222]: I'll add them tomorrow, thanks for the useful remarks :)

**Comment by jbartosik** [comment]
Please ping me when I should take another look

File: vertical-pod-autoscaler/pkg/recommender/input/cluster_feeder_test.go
Comment (matthyx) [inline vertical-pod-autoscaler/pkg/recommender/input/cluster_feeder_test.go:315]: @jbartosik I don't understand why the test fails here...
```
    cluster_feeder_test.go:315: missing call(s) to *mocktarget.MockVpaTargetSelectorFetcher.Fetch(is equal to &{{ } {testVpa  testNamespace    0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[] map[] [] []  []} {&CrossVersionObjectReference{Kind:dodokind,Name:dotaro,APIVersion:stardust,} <nil> 0xc000111098 []} {0xc0001110b0 []}}) /home/mbertschy/github.com/matthyx/autoscaler/vertical-pod-autoscaler/pkg/target/mock/fetcher_mock.go:57
    cluster_feeder_test.go:315: aborting test due to missing call(s)
```

**Comment by matthyx** [comment]
> Please ping me when I should take another look

I have added tests, please check their integration... and I also have an issue when I require early abort.

File: vertical-pod-autoscaler/pkg/recommender/input/cluster_feeder_test.go
Comment (jbartosik) [reply vertical-pod-autoscaler/pkg/recommender/input/cluster_feeder_test.go:315]: This is because all test cases here set up expectation that they'll Fetch VPA in [line 305](**https://github.com/kubernetes/autoscaler/pull/4799/files#diff-65acf6f86a7495618a847b01cbb7d4f26601545f611fd6ef5702326aeec03341R305):

```
targetSelectorFetcher.EXPECT().Fetch(vpa).Return(tc.selector, tc.fetchSelectorError)
```

and some test cases you add don't fetch VPA object (because they recognize this recommenders issn't supposed to handle the VPA object).

I think the easiest way to handle this is to add a bool to tell each test case if it should set up the expectation, like 

https://github.com/jbartosik/autoscaler/commit/024e1ce8a1fd392ddbb2ab7a4b9d446f2b7cf4f2

**Review by jbartosik — CHANGES_REQUESTED** [review]

**Comment by matthyx** [comment]
Thanks @jbartosik I've missed that... does my change suit you?

File: vertical-pod-autoscaler/pkg/recommender/input/cluster_feeder.go
Comment (jbartosik) [reply vertical-pod-autoscaler/pkg/recommender/input/cluster_feeder.go:364]: Ack.

**Review by jbartosik — CHANGES_REQUESTED** [review]
I think we're getting close. I left a few comments but I saw just some small stuff to change.

File: vertical-pod-autoscaler/pkg/recommender/input/cluster_feeder.go
Comment (matthyx) [reply vertical-pod-autoscaler/pkg/recommender/input/cluster_feeder.go:61]: cheers, not my mother tongue

**Comment by matthyx** [comment]
@jbartosik thanks for your help, I think we're good to go :)

**Review by jbartosik — APPROVED** [review]
Looks good to me. But I don't want to merge it for a while more. I want to finish releasing VPA 0.12.0 (I think code is good but e2e tests are timing out) before merging any changes to VPA that are not necessary for the release.

So

/hold

for now. I plan to unhold and merge after I publish the release

**Comment by matthyx** [comment]
@jbartosik can you also change the "Draft" status of the PR? I cannot do it, strangely :(

**Comment by jbartosik** [comment]
@matthyx can you add some documentation on running multiple recommenders. 

I think it can be quite simple, something like:
 
 - a yaml with recommender with different cpu percentile and name in flag
 - a short .md file pointing to the yaml and highlighting the flags and what they do.

**Pull request merged**
